src/main/java/org/folio/auth/authtokenmodule/storage/RefreshTokenStore.java

```diff
@@ -0,0 +1,199 @@
+
+    // Attempt to update the token to be revoked. If this update succeeds, the token
+    // will now be marked as revoked, and a single row will be returned. If no rows
+    // are returned, the update didn't succeed. This token has redeemed before
+    // and is therefore revoked.
+    String sql = "UPDATE " + table + "SET is_revoked=TRUE WHERE id=$1 AND is_revoked=FALSE";
+    Tuple params = Tuple.of(tokenId);
+
+    return pool.withConnection(conn -> {
+      return conn.preparedQuery(sql).execute(params).compose(rows -> {
+
+        // If the update has succeeded the row count will be 1. This means that the token
+        // was never redeemed. The update however has now made it revoked
+        // in an atomic operation. We can return success.
+        if (rows.rowCount() == 1) {
+          log.debug("Token {} has not yet been used, but is now revoked because it has been",
+            tokenId);
+          return Future.succeededFuture();
+        }
+
+        // The row count is not 1 so the token has been used more than once or has been
+        // revoked when another token has been used more than once.
+        String leakedMessage = "Refresh token {} is revoked." +
+          " Revoking all tokens for user {}.";
+        log.info(leakedMessage, tokenId, userId);
+
+        return revokeAllTokensForUser(conn, userId).compose(y -> {
+          return Future.failedFuture("Token leaked. All tokens for user are now revoked.");
+        });
+      });
+    });
+  }
+
+  private boolean tokenHasExpired(RefreshToken rt) {
+    return Instant.now().getEpochSecond() >= rt.getExpiresAt();
+  }
+
+  private Future<Void> revokeAllTokensForUser(SqlConnection conn, UUID userId) {
+    log.debug("Revoking all refresh tokens for user id {}", userId);
+
+    // First attempt to update but skipping any locks to be the most aggressve as we can.
+    String updateSkipLocked = "UPDATE " + tableName(tenant, REFRESH_TOKEN_SUFFIX) +
+      "SET is_revoked=TRUE " +
+      "WHERE id IN (SELECT id FROM " + tableName(tenant, REFRESH_TOKEN_SUFFIX) +
+                   "WHERE user_id=$1 AND is_revoked=FALSE " +
+                   "FOR UPDATE SKIP LOCKED)";
+
+     // Next come back and wait on any that were locked since we can't neglect them.
+     String updateAll = "UPDATE " + tableName(tenant, REFRESH_TOKEN_SUFFIX) +
+       "SET is_revoked=TRUE " +
+       "WHERE user_id=$1 AND is_revoked=FALSE";
+
+    Tuple params = Tuple.of(userId);
+
+    return conn.preparedQuery(updateSkipLocked).execute(params)
+      .compose(x -> conn.preparedQuery(updateAll).execute(params)).mapEmpty();
+  }
+
+  /**
+   * Cleans up (deletes) refresh tokens which have passed their time of expiration.
+   * Clients need to ensure that they are requesting new refresh tokens before they
+   * expire.
+   * @return A failed future should the deletion fail. A succeeded future if the
+   * deletion succeeds.
```

Comment:
This comment is not needed. This behavior is the default for any `Future<Void>`.

Reply:
Removed